### PR TITLE
on_evaluation_end for callback

### PR DESCRIPTION
**Pull request by joelthchao**
PR for #2555, #2548 and #2521.
Add `on_evaluate_end` to `Callback` to fill the missing step and improve customized callback flexibility.

`on_train_begin`
----`on_epoch_begin`
---------`on_batch_begin`
---------`on_batch_end`
----`on_epoch_begin`
---- (evaluation)
----`on_evaluate_end` (this PR)
`on_train_end`

Here is an example:
```python
x_train, y_train = np.random.rand(1024, 10), np.random.rand(1024, 1)
x_val, y_val = np.random.rand(512, 10), np.random.rand(512, 1)

net_input = Input(shape=(10,))
net = Dense(1)(net_input)

model = Model(net_input, net)

class EvaluateEndCallback(Callback):
    def on_evaluate_end(self, epoch, logs=None):
        print('epoch: {}, logs: {}'.format(epoch, logs))

model.compile(optimizer='sgd', loss='mse')
model.fit(x_train, y_train, validation_data=(x_val, y_val), callbacks=[EvaluateEndCallback()])
```
Using `on_epoch_end` (progress bar boken):
```
Epoch 1/10
 928/1024 [==========================>...] - ETA: 0s - loss: 0.2469epoch: 0, logs: {'val_loss': 0.20372078195214272, 'loss': 0.24096788465976715}

1024/1024 [==============================] - 0s - loss: 0.2410 - val_loss: 0.2037
```
Using `on_evaluate_end`:
```
Epoch 1/10
1024/1024 [==============================] - 0s - loss: 0.3038 - val_loss: 0.2313
epoch: 0, logs: {'val_loss': 0.23126391600817442, 'loss': 0.30380106996744871}
```

### CODE REVIEW

**Comment by bstriner** [comment]
It seems like your issue is that the progress callback is always added to the end of the list of callbacks, but you want it to be called before your callback.

Is your issue fixed with existing keras if you do something like this?

```python
callbacks = [ProgbarLogger(), EvaluateEndCallback()]
model.fit(...,verbose=0)
```

Cheers

**Comment by bstriner** [comment]
@joelthchao actually, you would need to do something like this to hide the default progbar. Please let me know if this works for you.

```python
class ProgbarLoggerVerbose(ProgbarLogger):
    def on_train_begin(self, logs=None):
        ProgbarLogger.on_train_begin(self, logs)
        self.verbose = True
callbacks = [ProgbarLoggerVerbose(), EvaluateEndCallback()]
model.fit(...,callbacks =callbacks, verbose=0)
```

**Comment by joelthchao** [comment]
@bstriner Thanks for your comments!
> It seems like your issue is that the progress callback is always added to the end of the list of callbacks, but you want it to be called before your callback.

This is one of my purposes, but most important thing is to fill the missing step of callback.

> actually, you would need to do something like this to hide the default progbar. Please let me know if this works for you.

This trick will work. For me, I would prefer a native way to do things after evaluation, without do little hack to hide progbar. For example, stats of predicted classes, value monitoring, on_evaluate_end would help a lot.

File: keras/engine/training.py [hunk 1]
```diff
@@ -910,6 +910,8 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=32,
                         for l, o in zip(out_labels, val_outs):
                             epoch_logs['val_' + l] = o
             callbacks.on_epoch_end(epoch, epoch_logs)
+            if do_validation:
+                callbacks.on_evaluate_end(epoch, epoch_logs)
```
Comment (fchollet) [inline keras/engine/training.py]: How is there any difference between this and `on_epoch_end`? In practice, both are called in succession, with nothing in between.

**Comment by joelthchao** [comment]
@fchollet They are called in succession when "only one" callback exists.
When verbose is on, `ProgbarLogger` callback do the progbar update in `on_epoch_end` and ProgbarLogger is added to the end of callbacks by default.
```python
callbacks = [cbks.BaseLogger()] + (callbacks or []) + [self.history]
        if verbose:
            callbacks += [cbks.ProgbarLogger()]
```
So the order of consecution is:
`yourcallback.on_epoch_end` -> `progbar.on_epoch_end` -> `yourcallback.on_evaluate_end`

**Comment by fchollet** [comment]
Concretely, what use cases do you envision for this method?

**Comment by joelthchao** [comment]
@fchollet Here are some my own usages:
1. Print out stats of validation: distribution of predicted class, fail cases analysis
2. Debug info: value of some weights in the model, value of feature vector from some layers
If you feel this is unnecessary to appear in Keras, I am fine to close this PR.

**Comment by fchollet** [comment]
Sorry to take time to make a decision on this. Overall, it seems this is too niche to be included in core Keras. Since no one else has been requesting this feature (including in this PR, which has been around for some time), we won't be moving forward with it.

Again, sorry about the delay.

**Comment by bstriner** [comment]
I have been considering trying to add callbacks during evaluation and prediction. Would like to add pretty progress bars to `keras-tqdm` for evaluation and prediction as well. That might incidentally give @joelthchao what he is looking for as well.

**Comment by joelthchao** [comment]
@bstriner 
> Please let me know if this works for you.

When using `fit_generator`, it needs to change to
```python
callbacks = [ProgbarLoggerVerbose('steps'), EvaluateEndCallback()]
```

File: keras/engine/training.py (hunk 1 above)
Comment (Cerno-b) [inline keras/engine/training.py]: Sorry for being obtuse, but I still see some sense in having on_evaluation_end(). 

In my specific case, I am trying to access the model history in a custom callback to write some graphs for internal states of the current epoch. It seems to me that keras updates the history after my callback is called, so when I try to access history.history["loss"] I actually get the loss of the epoch before the current one.

I admit, I don't know the best solution here. It certainly has merit to have the history updated after all other callbacks. On the other hand, adding an on_evaluate_end just for the history update may be a bit too much. Adding a flag somewhere that pushes a callback's execution to the very end may be an option, but I'm too fresh on the concepts of keras so I don't know what would be the cleanest solution here.
